source/common/buffer/buffer_impl.h

```diff
@@ -10,13 +10,7 @@
 namespace Envoy {
 namespace Buffer {
 
-class OwnedImplFactory : public Factory {
-public:
-  // Buffer::Factory
-  InstancePtr create() override;
-};
-
-class LibEventInstance : public Instance {
+class LibEventInstance : public virtual Instance {
```

Comment:
Random thought: What if we got rid of LibEventInstance entirely and just had OwnedImpl and WatermarkBuffer both inherent from Instance? They could both use a common helper class/mix-in to implement most of the functionality? This would reduce some of the diamond/virtual nastiness at the expense of slightly more boilerplate?

Reply:
I guess you need some base class because of the move stuff. I'm brainstorming on the following:

Let's say we had a mix-in base class called `BufferBase`. OwnedImpl and WatermarkBuffer both inherit from `BufferBase` as well as `Instance` and call into `BufferBase` where needed. I think in this case all the move stuff could still happen via static_cast, and we would remove some of the virtual hacks, post process, etc.

Reply:
I think that could work but that'd mean that BufferBase basically implemented all the Instance functions with slightly different names, which seems like a lot of code duplication to avoid a dynamic cast.   If that's our only savings I think I prefer this version to that.

I don't think anything would remove postProcess because we have to handle the case where some other actor grabs the underlying data via buffer() and alters the data.  There has to be a follow-up call to give the watermark buffer a chance to act, no?  Note we're not calling postProcess() in move, we're calling rhs.postProcess() - allowing the "external" buffer to do work.  We could alternately replace the two calls to buffer().get() with wrapper functions, but we can do that as easily in the current LibEventInstance (which I'd be happy to do).

Reply:
I guess my thinking here is that this code is called a ton. The dynamic casts will add up. Normally I would agree, but I wonder if in this case it is actually worth optimizing for static_cast vs. dynamic_cast. I don't feel that strongly. @htuch any opinion?

Reply:
If this is about `dynamic_cast` overhead, why not just use `static_cast` in release builds and do some `ASSERT` with `dynamic_cast` for debug/fastbuilds?

Reply:
I think because of virtual inheritance you need to use static_cast for it to work correctly, right @alyssawilk? In this particular case, it doesn't bother me that much to get rid of the complex virtual inheritance, and forward the calls to the base class (with slightly different names), and the in the case of watermark, additionally do the check functions (I would note that in watermark code you are already basically doing this, so LOC difference would be small vs. what I'm proposing).

Reply:
That still doesn't work though.  Assuming we're talking
class BufferBase { // buffer functionality}
class OwnedImpl : public BufferBase, public Instance { //calls wrapper functions }
class WatermarkBuffer : public BufferBase, public WatermarkInstance { //calls wrapper functions }

Then for the OwnedImpl:Move(Instance& rhs) we can't cast to BufferBase because of the inheritance model and casting to OwnedImpl won't work if it's a WatermarkBuffer.

If we're really worried about perf, I think the simplest thing to do is revert back to the initial commit (modulo any relevant comments) 
https://github.com/lyft/envoy/pull/1562/commits/dbd5935d8f4fae27aedc495fe1503a629c4c764f

As said I wasn't crazy about the static cast in source/common/network/connection_impl.cc
but I don't think it's terrible and it avoids all the inheritance complexity.  WDYT?


https://github.com/lyft/envoy/pull/1562/commits/dbd5935d8f4fae27aedc495fe1503a629c4c764f

Reply:
Right OK makes sense, sorry, didn't think this all the way though. FWIW I think I like https://github.com/lyft/envoy/commit/dbd5935d8f4fae27aedc495fe1503a629c4c764f better than the alternative in general, even apart from dynamic_casts. Which do you prefer?

If you are OK going back to that let's refresh on that one and I think my other comments apply:
1) Make read_buffer_ inline in ConnectionImpl.
2) Potentially not use templates in mocks, but will leave that up to you.

Reply:
I reverted via diff patch so basically cherrypicked out the change we wanted to discard.  I'd prefer to leave the templates in mocks as it shouldn't be hard to maintain and read_buffer_ is inline :-)

Reply:
Any other requests on this one?

Reply:
No I don't think so. (I assume you mean before you make the above changes?)

Reply:
PEBKAC.  I failed to notice my last push bounced due to the partial rollback.  Pushed, and should be as discussed, with read_buffer_ inline and no dynamic casts.
Sorry for the push -f killing outstanding state but the diffs were nearly 100% in any case!